File: facade/src/main/java/org/tdf/sunflower/vm/WBI.java
package org.tdf.sunflower.vm;

import lombok.Value;
import org.tdf.common.types.Uint256;
import org.tdf.common.util.BigIntegers;
import org.tdf.common.util.FastByteComparisons;
import org.tdf.common.util.HexBytes;
import org.tdf.lotusvm.ModuleInstance;
import org.tdf.lotusvm.types.CustomSection;
import org.tdf.lotusvm.types.Module;
import org.tdf.sunflower.vm.abi.Abi;
import org.tdf.sunflower.vm.abi.WbiType;

import java.io.ByteArrayOutputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public abstract class WBI {

    @Value
    public static class InjectResult {
        String function;
        long[] pointers;
        boolean executable;
    }

    public static byte[] dropInit(byte[] code) {
        Module m = new Module(code);
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        // drop __init sections
        int now = 0;
        for (CustomSection section : m.getCustomSections()) {
            if (section.getName().equals("__init")) {
                out.write(code, now, section.getOffset() - now);
                now = section.getLimit();
            }
        }
        out.write(code, now, code.length - now);
        return out.toByteArray();
    }

    public static byte[] extractInitData(Module m) {
        return m.getCustomSections().stream().filter(x -> x.getName().equals("__init")).findFirst()
            .map(CustomSection::getData).orElse(new byte[0]);
    }

    // the __init section is dropped before inject
    public static InjectResult inject(boolean create, Abi abi, ModuleInstance i, HexBytes input) {
        List<Abi.Entry.Param> params = null;
        HexBytes encoded = null;
        String function = null;

        // 2. for contract create, constructor is not necessarily
        if (create && abi.findConstructor() != null) {
            function = "init";
            params = abi.findConstructor().inputs;
            encoded = input;
        }

        // 3. for contract call, find function by signature
        if (!create) {
            Abi.Function f = abi.findFunction(x -> FastByteComparisons.equal(x.encodeSignature(), Arrays.copyOfRange(input.getBytes(), 0, 4)));
            Objects.requireNonNull(f);
            function = f.name;
            params = f.inputs;
            Objects.requireNonNull(params);
            encoded = input.slice(4);
        }

        if (params == null)
            return new InjectResult(function, new long[0], false);

        // malloc param types
        List<?> inputs = Abi.Entry.Param.decodeList(params, encoded.getBytes());
        long[] ret = new long[params.size()];

        for (int j = 0; j < inputs.size(); j++) {
            Abi.Entry.Param p = params.get(j);

            switch (p.type.getName()) {
                case "uint8":
                case "uint16":
                case "uint32":
                case "uint64": {
                    ret[j] = ((BigInteger) inputs.get(j)).longValueExact();
                    break;
                }
                case "uint":
                case "uint256": {
                    BigInteger b = (BigInteger) inputs.get(j);
                    ret[j] = WBI.malloc(i, Uint256.of(b));
                    break;
                }
                case "string": {
                    String s = (String) inputs.get(j);
                    ret[j] = WBI.malloc(i, s);
                    break;
                }
                case "address": {
                    byte[] addr = (byte[]) inputs.get(j);
                    ret[j] = WBI.mallocAddress(i, HexBytes.fromBytes(addr));
                    break;
                }
                default: {
                    if (p.type.getName().endsWith("]") || p.type.getName().endsWith(")")) {
                        throw new RuntimeException("array or tuple is not supported");
                    }

                    if (p.type.getName().startsWith("bytes")) {
                        byte[] data = (byte[]) inputs.get(j);
                        ret[j] = WBI.mallocBytes(i, HexBytes.fromBytes(data));
                    }
                    break;
                }
            }
        }
        return new InjectResult(function, ret, true);
    }

    // String / U256 / HexBytes
    public static Object peek(ModuleInstance instance, int offset, int type) {
        long t = Integer.toUnsignedLong(type);
        long startAndLen = instance.execute("__peek", offset, t)[0];
        int start = (int) (startAndLen >>> 32);
        int len = (int) (startAndLen);
        byte[] bin = instance.getMemory().loadN(start, len);
        switch (type) {
            case WbiType.STRING: {
                return new String(bin, StandardCharsets.UTF_8);
            }
            case WbiType.UINT_256: {
                return Uint256.of(bin);
            }
            case WbiType.BYTES:
            case WbiType.ADDRESS: {
                return HexBytes.fromBytes(bin);
            }
        }
        throw new RuntimeException("unexpected");
    }

    private static int mallocInternal(ModuleInstance instance, int type, byte[] bin) {
        long t = Integer.toUnsignedLong(type);
        long ptr = instance.execute("__malloc", bin.length)[0];
        instance.getMemory().put((int) ptr, bin);
        long p = instance.execute("__change_t", t, ptr, bin.length)[0];
        int r = (int) p;
        if (r < 0)
            throw new RuntimeException("malloc failed: pointer is negative");
        return r;
    }


    public static int malloc(ModuleInstance instance, String s) {
        byte[] bin = s.getBytes(StandardCharsets.UTF_8);
        return mallocInternal(instance, WbiType.STRING, bin);
    }

    public static int malloc(ModuleInstance instance, Uint256 s) {
        byte[] bin = s.getNoLeadZeroesData();
        return mallocInternal(instance, WbiType.UINT_256, bin);
    }

    public static int mallocBytes(ModuleInstance instance, HexBytes bin) {
        return mallocInternal(instance, WbiType.BYTES, bin.getBytes());
    }

    public static int mallocAddress(ModuleInstance instance, HexBytes address) {
        return mallocInternal(instance, WbiType.ADDRESS, address.getBytes());
    }
}

File: facade/src/main/java/org/tdf/sunflower/vm/WBI.kt
package org.tdf.sunflower.vm

import org.tdf.common.types.Uint256
import org.tdf.common.util.*
import org.tdf.lotusvm.Module
import org.tdf.lotusvm.ModuleInstance
import org.tdf.sunflower.vm.abi.Abi
import org.tdf.sunflower.vm.abi.WbiType
import java.math.BigInteger
import java.nio.charset.StandardCharsets

/**
 * TODO:
 * 1. instead of json, use custom sections named with __abi like foo(address,address)returns(uint256)
 * ident=[a-zA-Z_][a-zA-Z0-9_]{0,127}
 * type = (address|uintx|intx|bytes|string)
 * $ident\((($type,)*$type)?\)(returns\($type\))?
 * 2. use wasmer instead of lotusvm
 */
object WBI {
    const val ABI_SECTION_NAME = "__abi";
    const val INIT_SECTION_NAME = "__init"
    const val WBI_MALLOC = "__malloc"
    const val WBI_PEEK = "__peek"
    const val WBI_CHANGE_TYPE = "__change_t"
    const val WBI_MALLOC_256 = "__malloc_256"
    const val WBI_MALLOC_512 = "__malloc_512"

    val REVERSED = setOf(WBI_MALLOC, WBI_PEEK, WBI_CHANGE_TYPE, WBI_MALLOC_256, WBI_MALLOC_512)


    fun dropInit(code: ByteArray): ByteArray {
        return code;
    }

    fun extractInitData(m: Module): ByteArray {
        return m.customSections.firstOrNull { it.name == INIT_SECTION_NAME }?.data ?: ByteUtil.EMPTY_BYTE_ARRAY
    }

    // the __init section is dropped before inject
    fun inject(create: Boolean, abi: Abi, i: ModuleInstance, input: HexBytes): InjectResult {
        var function: String? = null
        var entry: Abi.Entry? = null
        var encoded: HexBytes? = null

        // 2. for contract create, constructor is not necessarily
        if (create && abi.findConstructor() != null) {
            entry = abi.findConstructor()
            function = "init"
            encoded = input
        }

        // 3. for contract call, find function by signature
        if (!create) {
            val sig = input.bytes.selector()
            val f = abi.findFunction { x: Abi.Function -> x.encodeSignature().contentEquals(sig) }!!
            entry = f
            function = f.name
            // drop signature parts
            encoded = input.bytes.unselect().hex()
        }

        // params == null -> abi not found
        if (entry == null)
            return InjectResult(function ?: "", entry, LongArray(0), false)

        // malloc param types
        val inputs = Abi.Entry.Param.decodeList(entry.inputs, encoded!!.bytes)
        val ret = LongArray(entry.inputs.size)
        for (j in inputs.indices) {
            val p = entry.inputs[j]
            when (p.type.name) {
                "uint8", "uint16", "uint32", "uint64" -> {
                    ret[j] = (inputs[j] as BigInteger).longValueExact()
                }
                "uint", "uint256" -> {
                    val b = inputs[j] as BigInteger
                    ret[j] = malloc(i, Uint256.of(b)).toLong()
                }
                "string" -> {
                    val s = inputs[j] as String
                    ret[j] = malloc(i, s).toLong()
                }
                "address" -> {
                    val addr = inputs[j] as ByteArray
                    ret[j] = mallocAddress(i, addr.hex()).toLong()
                }
                else -> {
                    if (p.type.name.endsWith("]") || p.type.name.endsWith(")")) {
                        throw RuntimeException("array or tuple is not supported yet")
                    }
                    if (p.type.name.startsWith("bytes")) {
                        val data = inputs[j] as ByteArray
                        ret[j] = mallocBytes(i, data.hex()).toLong()
                    }
                }
            }
        }
        return InjectResult(function!!, entry, ret, true)
    }

    // String / U256 / HexBytes
    @JvmStatic
    fun peek(instance: ModuleInstance, offset: Int, type: Long): Any {
        val startAndLen = instance.execute(WBI_PEEK, offset.toLong(), type)[0]
        val start = (startAndLen ushr 32).toInt()
        val len = startAndLen.toInt()
        val bin = ByteArray(len)
        instance.memory.read(start, bin)
        when (type) {
            WbiType.STRING -> {
                return String(bin, StandardCharsets.UTF_8)
            }
            WbiType.UINT_256 -> {
                return Uint256.of(bin)
            }
            WbiType.BYTES, WbiType.ADDRESS, WbiType.BYTES_32 -> {
                return bin.hex()
            }
        }
        throw RuntimeException("unexpected")
    }

    fun mallocWords(instance: ModuleInstance, data: LongArray): Int {
        val p =
        when(data.size) {
            4 -> instance.execute(WBI_MALLOC_256, *data)[0]
            8 -> instance.execute(WBI_MALLOC_512, *data)[0]
            else -> throw RuntimeException("invalid size")
        }
        if(p > Int.MAX_VALUE)
            throw RuntimeException("malloc failed: pointer is negative")
        return p.toInt()
    }

    fun malloc(instance: ModuleInstance, type: Long, bin: ByteArray): Int {
        val ptr = instance.execute(WBI_MALLOC, bin.size.toLong())[0]
        instance.memory.write(ptr.toInt(), bin)
        val p = instance.execute(WBI_CHANGE_TYPE, type, ptr, bin.size.toLong())[0]
        val r = p.toInt()
        if (r < 0) throw RuntimeException("malloc failed: pointer is negative")
        return r
    }

    fun malloc(instance: ModuleInstance, s: String): Int {
        val bin = s.toByteArray(StandardCharsets.UTF_8)
        return malloc(instance, WbiType.STRING, bin)
    }

    fun malloc(instance: ModuleInstance, s: Uint256): Int {
        val bin = s.byte32
        val data = LongArray(4)
        data[0] = BigEndian.decodeInt64(bin, 0)
        data[1] = BigEndian.decodeInt64(bin, 8)
        data[2] = BigEndian.decodeInt64(bin, 16)
        data[3] = BigEndian.decodeInt64(bin, 24)
        return mallocWords(instance, data)
    }

    fun mallocBytes(instance: ModuleInstance, bin: HexBytes): Int {
        return malloc(instance, WbiType.BYTES, bin.bytes)
    }

    fun mallocAddress(instance: ModuleInstance, address: HexBytes): Int {
        return malloc(instance, WbiType.ADDRESS, address.bytes)
    }

    class InjectResult(val name: String, val entry: Abi.Entry?, val pointers: LongArray, val executable: Boolean);
}